core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala

```diff
@@ -196,6 +201,110 @@ class ReplicaAlterLogDirsThreadTest {
+    when(replicaManager.onlinePartition(t1p0)).thenReturn(Some(partition))
+    when(replicaManager.getPartitionOrException(t1p0)).thenReturn(partition)
+    when(replicaManager.logManager).thenReturn(logManager)
+    doAnswer(_ => {
+      logCleaner.abortAndPauseCleaning(t1p0)
+    }).when(logManager).abortAndPauseCleaning(t1p0)
+    doAnswer(_ => {
+      logCleaner.resumeCleaning(Seq(t1p0))
+    }).when(logManager).resumeCleaning(t1p0)
+
+    when(quotaManager.isQuotaExceeded).thenReturn(false)
+
+    when(partition.lastOffsetForLeaderEpoch(Optional.empty(), leaderEpoch, fetchOnlyFromLeader = false))
+      .thenReturn(new EpochEndOffset()
+        .setPartition(partitionId)
+        .setErrorCode(Errors.NONE.code)
+        .setLeaderEpoch(leaderEpoch)
+        .setEndOffset(logEndOffset))
+    when(partition.futureLocalLogOrException).thenReturn(futureLog)
+    doNothing().when(partition).truncateTo(offset = 0, isFuture = true)
+    when(partition.maybeReplaceCurrentWithFutureReplica()).thenReturn(false)
+
+    when(futureLog.logStartOffset).thenReturn(0L)
+    when(futureLog.logEndOffset).thenReturn(0L)
+    when(futureLog.latestEpoch).thenReturn(None)
+
+    val requestData = new FetchRequest.PartitionData(topicId, 0L, 0L,
+      config.replicaFetchMaxBytes, Optional.of(leaderEpoch))
+    val responseData = new FetchPartitionData(
+      Errors.NONE,
+      0L,
+      0L,
+      MemoryRecords.withRecords(CompressionType.NONE, new SimpleRecord(1000, "foo".getBytes(StandardCharsets.UTF_8))),
+      Optional.empty(),
+      OptionalLong.empty(),
+      Optional.empty(),
+      OptionalInt.empty(),
+      false)
+    mockFetchFromCurrentLog(tid1p0, requestData, config, replicaManager, responseData)
+
+    val endPoint = new BrokerEndPoint(0, "localhost", 1000)
+    val leader = new LocalLeaderEndPoint(endPoint, config, replicaManager, quotaManager)
+    val thread = new ReplicaAlterLogDirsThread(
+      "alter-logs-dirs-thread",
+      leader,
+      failedPartitions,
+      replicaManager,
+      quotaManager,
+      new BrokerTopicStats,
+      config.replicaFetchBackoffMs)
+
+    // before starting the fetch, pause the clean of the partition.
+    logManager.abortAndPauseCleaning(t1p0)
+    thread.addPartitions(Map(t1p0 -> initialFetchState(fetchOffset = 0L, leaderEpoch)))
+    assertTrue(thread.fetchState(t1p0).isDefined)
+    assertEquals(1, thread.partitionCount)
+
+    // first test: get handle without exception
+    when(partition.appendRecordsToFollowerOrFutureReplica(any(), ArgumentMatchers.eq(true))).thenReturn(None)
+
+    thread.doWork()
+
+    assertTrue(thread.fetchState(t1p0).isDefined)
+    assertEquals(1, thread.partitionCount)
+    assertTrue(logCleaner.isCleaningInStatePaused(t1p0))
+
+    // second test: process partition data with throwing a KafkaStorageException.
+    when(partition.appendRecordsToFollowerOrFutureReplica(any(), ArgumentMatchers.eq(true))).thenThrow(new KafkaStorageException("disk error"))
```

Comment:
Do we have a way to verify the log is cleaned up after it is in failed state?

Reply:
This is a good question - right now the test checks that the state has changed from paused to cleanable, but it does not check that the cleaning was successful. When you ask whether the state can be resumed correctly do you mean that cleaning can be resumed correctly (original partition) or whether the partition which tried to copy things can resume copying data correctly (replica partition)? If it is the first, then for the sake of the unit test isn't the current assertion that we are no longer in the cleaning state enough? If it is the second, I don't think that we expect the ReplicaFetchDirThread to recover from a KafkaStorageException without a broker reboot.

Reply:
My question is for the replica partition. How do we recover (or ensure we don't recover) that one.